File: src/y17.rb
p ->() {

  # Fix Points
  # Higher Order Functions
  # Refactorings
  # - Inline
  # - Tennent Correspondence Principle
  # Factorial is the fixpoint of FIMP

  # ->(n) { n.zero? ? 1 : n * fact.(n-1) }.(5)

  ->(f) {
    ->(g) { ->(n) { f.(g.(g)).(n) } }.(
      ->(g) { ->(n) { f.(g.(g)).(n) } } )
  }.(->(f) {
      ->(n) { n.zero? ? 1 : n * f.(n-1) }
    }).(5)

}.()

File: src/y17.clj
(println
 (do
   ;; Fixpoint
   ;; Higher Order Functions
   ;; Functional Refactorings
   ;; (1) Tennent Correspondence Principle
   ;; (2) Introduce Binding
   ;; (3) Wrap Function
   ;; (4) Inline Defintion
   ;; Recursion

   (def error (fn [n] (throw (Throwable. "SHOULD NOT BE CALLED"))))

   (def fact-improver
        (fn [partial]
          (fn [n] (if (zero? n) 1 (* n (partial (dec n)))))))

   ;; Fixpoint combinator
   ;; Applicative Order Y Combinator (Z-Combinator)
   (def y
        (fn [f]
          ((fn [x] (f (fn [v] ((x x) v))))
           (fn [x] (f (fn [v] ((x x) v)))))))

   ;; Y calculates the fixpoint of an improver function
   (def fact (y fact-improver))

   ;; fact is the fixpoint of the fact-improver function
   (def fact (fact-improver fact))

   (fact 5)

))